cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java

```diff
@@ -0,0 +1,137 @@
+import co.cask.cdap.common.lang.InstantiatorFactory;
+import co.cask.cdap.common.lang.PropertyFieldSetter;
+import co.cask.cdap.internal.app.runtime.DataSetFieldSetter;
+import co.cask.cdap.internal.app.runtime.MetricsFieldSetter;
+import co.cask.cdap.internal.lang.Reflections;
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.tephra.TransactionContext;
+import co.cask.tephra.TransactionFailureException;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.reflect.TypeToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Execute the custom action in the Workflow.
+ */
+public class CustomActionExecutor {
+  private static final Logger LOG = LoggerFactory.getLogger(CustomActionExecutor.class);
+  private final ProgramRunId workflowRunId;
+  private final BasicWorkflowContext workflowContext;
+  private final WorkflowAction action;
+
+  /**
+   * Creates instance which will be used to initialize, run, and destroy the custom action.
+   * @param workflowRunId the Workflow run which started the execution of this custom action.
+   * @param workflowContext an instance of context
+   * @param instantiator to instantiates the custom action class
+   * @param classLoader used to load the custom action class
+   * @throws Exception when failed to instantiate the custom action
+   */
+  public CustomActionExecutor(ProgramRunId workflowRunId, BasicWorkflowContext workflowContext,
+                              InstantiatorFactory instantiator, ClassLoader classLoader) throws Exception {
+    this.workflowRunId = workflowRunId;
+    this.workflowContext = workflowContext;
+    this.action = createAction(workflowContext, instantiator, classLoader);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFactory instantiator,
+                                      ClassLoader classLoader) throws Exception {
+    Class<?> clz = Class.forName(context.getSpecification().getClassName(), true, classLoader);
+    Preconditions.checkArgument(WorkflowAction.class.isAssignableFrom(clz), "%s is not a WorkflowAction.", clz);
+    WorkflowAction action = instantiator.get(TypeToken.of((Class<? extends WorkflowAction>) clz)).create();
+    Reflections.visit(action, action.getClass(),
+                      new PropertyFieldSetter(context.getSpecification().getProperties()),
+                      new DataSetFieldSetter(context),
+                      new MetricsFieldSetter(context.getMetrics()));
+    return action;
+  }
+
+  void execute() throws Exception {
+    ClassLoader oldClassLoader = setContextCombinedClassLoader(action);
+    try {
+      initializeInTransaction();
+      runInTransaction();
+    } catch (Throwable t) {
+      LOG.error("Error executing the action {} in the Workflow {}.", workflowContext.getSpecification().getName(),
```

Comment:
If you propagate an exception, then no need to log. It should be the workflow driver who log it.